src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1063 @@
+
+			for(int line = 0; line < 14; line++)
+			{
+				// This is at the start of a 9-dot raster-line segment in a single character
+				uint8_t data = 0x00;
+				int dot_nr8 = 0;
+				int extra_intensity = 1;
+				if(in_line_char_nr >= 0 && in_line_char_nr < 80)
+				{
+					bool const draw_cursor = (is_cursor && (line == 12 || block_cursor) && (!blinking_cursor || cursor_blink_strobe));
+					int const vram_data = m_vram[vram_address];
+
+					// Pattern generator & character blanking
+					if(m_video_enable && (vram_data&0x60 || !blank_ctrl_chars) && !(vram_data&0x80 && blank_attr_chars))
+					{
+						data = m_font[((vram_data&0x7f)<<4) + line];
+						if(extend_dot7)
+						{
+							dot_nr8 = data&0x01;
+						}
+					}
+
+					// Pattern modifiers
+					if(underline_mode)
+					{
+						// Note: The underline in underline-mode is not affected by blanking!
+						if((vram_data&0x80) && line == 13)
+						{
+							data = 0xff;
+							dot_nr8 = 0x01;
+						}
+					}
+					else if(m_video_enable) // attribute mode
+					{
+						if(vram_data&0x80)
+						{
+							attribute = (vram_data&0x70)>>4;
+						}
+						else
+						{
+							if(attribute == 2)
+							{
+								// Low intensity
+								extra_intensity = 0;
+							}
+							else if(attribute == 3 && blink_strobe && !draw_cursor)
+							{
+								// Blink
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+							else if(attribute == 4)
+							{
+								// Inverse
+								data = ~data;
+								dot_nr8 = ~dot_nr8;
+							}
+							else if(attribute == 5 && line == 13)
+							{
+								// Underline
+								data = 0xff;
+								dot_nr8 = 0x01;
+							}
+							else if(attribute == 6)
+							{
+								// Invissible
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+						}
+					}
+
+					if(draw_cursor)
+					{
+						data = ~data;
+						dot_nr8 = ~dot_nr8;
+					}
+				}
+				if(full_inverse_video)
+				{
+					data = ~data;
+					dot_nr8 = ~dot_nr8;
+				}
+				for(int dot = 0; dot < 8; dot++)
+				{
+					bitmap.pix(chr_y_pos + line, chr_x_pos + dot) = palette[BIT(data, 7 - dot)<<extra_intensity];
+				}
+				bitmap.pix(chr_y_pos + line, chr_x_pos + 8) = palette[BIT(dot_nr8, 0)<<extra_intensity];
+			}
+		}
+	}
+	return 0;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// UART
+//
+
+void tandberg_tdv2100_disp_logic_device::update_rs232_lamps()
+{
+	m_write_waitl_cb(m_data_terminal_ready && !m_rx_handshake);
+	bool need_tx_for_on_line_led = ((m_dsw_u39->read()&0x0c0) == 0x080);
+	m_write_onlil_cb(m_data_terminal_ready && m_rx_handshake && (m_tx_handshake || !need_tx_for_on_line_led));
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_rxd_w(int state)
+{
+	if(!(m_data_terminal_ready && m_rx_handshake))
+	{
+		state = 1;
+	}
+	bool ext_echo = BIT(m_sw_rs232_settings->read(), 3);
+	if(!(ext_echo && m_data_terminal_ready))
+	{
+		state = (state && m_uart->so_r());
+	}
+	m_uart->write_si(state);
+}
```

Comment:
If `m_data_terminal_ready` or `m_rx_handshake` changes state or the external echo switch is flipped, shouldn’t that AY51013’s SI input be update immediately rather than waiting for the next state change of the external RxD input?

You’ll need to keep the input state so you can update the AY51013’s SI input when the other things that affect it change.

Reply:
Some of the logic in the RS232 handling has now been streamlined, to not duplicate the same responsibility across multiple functions. This should be sorted out by this change.